{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monday June 24, 2019\n",
    "Today, I met with Dr. Ussery, Kaleb, and Zulema and discussed my results. It seems good that the the program is finding around 2700 genes for 10 *E. coli* genomes, but Zulema thinks it might be low. The plan now is to proceed with testing the program for 20 genomes to see if the program is stable as the number of genomes increases. I will then compare the 20 genome runs to see if any of the criteria work better as the number of genomes increases. \n",
    "20 genomes: \n",
    "* GCA_000005845.2_ASM584v2.faa\n",
    "* GCA_000006665.1_ASM666v1.faa\n",
    "* GCA_000006925.2_ASM692v2.faa\n",
    "* GCA_000007445.1_ASM744v1.faa\n",
    "* GCA_000008865.2_ASM886v2.faa\n",
    "* GCA_000010245.1_ASM1024v1.faa\n",
    "* GCA_000010485.1_ASM1048v1.faa\n",
    "* GCA_000012005.1_ASM1200v1.faa\n",
    "* GCA_000012025.1_ASM1202v1.faa\n",
    "* GCA_000019645.1_ASM1964v1.faa\n",
    "* GCA_000026325.2_ASM2632v2.faa\n",
    "* GCA_000091005.1_ASM9100v1.faa\n",
    "* GCA_000092525.1_ASM9252v1.faa\n",
    "* GCA_000167875.1_ASM16787v1.faa\n",
    "* GCA_000176655.2_ASM17665v2.faa\n",
    "* GCA_000176695.2_ASM17669v2.faa\n",
    "* GCA_000333215.1_ASM33321v1.faa\n",
    "* GCA_000350785.1_Esch_coli_KTE21_V1.faa\n",
    "* GCA_000613265.1_ASM61326v1.faa\n",
    "* GCA_000690815.1_ASM69081v1.faa\n",
    "\n",
    "## Tuesday June 25, 2019\n",
    "I am continuing to test the program with 20 genomes today. I am having to rerun some of the results from yesterday as I think that I compared a different set of genomes than what I thought I was using. To fix this, I picked the 20 genomes that I thought I was testing and then I moved them to a separate demarcated folder (20_genomes). That way, I can direct the program to that specific folder and it can only read in the correct 20 genomes. \n",
    "\n",
    "## Wednesday June 26, 2019\n",
    "Today I continued getting data for 20 genome groupings under different conditions. I also **finally** got the pie chart labels fixed in FindMyFriends! I simply had to change the label positions to be subtracted from the total size of the pie chart! Code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Old version:\n",
    "labPos <- groups$nGenes/2 + cumsum(c(0, groups$nGenes[-nrow(groups)]))\n",
    "\n",
    "## Corrected version:\n",
    "labPos <- sum(groups$nGenes) - (groups$nGenes/2 + cumsum(c(0, groups$nGenes[-nrow(groups)])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(These are two different results, so the numbers are different, but the labels are better now!)\n",
    "\n",
    "Old: ![old](images/w6/20_genomes_4mer_top_9.png)\n",
    "\n",
    "Corrected: ![new](images/w6/20_genome_4mer_top_3_all_FYWH_CM.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From my runs, I was able to get a good amount of data about how well the criteria are working: ![gc](images/w6/grouping_comparisons.png)\n",
    "\n",
    "\n",
    "Once I added 10 more genomes to my analysis, my core genes dropped to around 2100. Kaleb took a look at the genomes that I was using, and it seems that the core genes might have dropped because my dataset was not a good representative dataset to compare between 10 and 20 genomes; the 20 genomes that I looked at were more diverse than the 10 genomes that I looked at since I did not pick as diverse of a set of 10 genomes. 10 of my 20 genomes were the original 10 genomes, but the 10 genomes that I added contained *E. coli* of a phylotype that was not present in the 10 genomes. To see if this is the case, Kaleb is going to send my a dataset that includes a representative genome from each of the 14 *E. coli* phylotypes. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For future reference: Here is a list of the 25 *E. coli* genomes that I have been using and their associated phylotypes:\n",
    "\n",
    "Assembly:\tPhylotype\n",
    "\n",
    "GCA_000092525.1\tShig2\n",
    "\n",
    "GCA_000006925.2\tShig1\n",
    "\n",
    "GCA_000006665.1\tE2\n",
    "\n",
    "GCA_000008865.2\tE2\n",
    "\n",
    "GCA_000019645.1\tF\n",
    "\n",
    "GCA_000176695.2\tG\n",
    "\n",
    "GCA_000012005.1\tE\n",
    "\n",
    "GCA_000176575.2\tE\n",
    "\n",
    "GCA_000350785.1\tD3\n",
    "\n",
    "GCA_000176655.2\tD2\n",
    "\n",
    "GCA_000026325.2\tD1\n",
    "\n",
    "GCA_000091005.1\tC\n",
    "\n",
    "GCA_000010485.1\tB2-1\n",
    "\n",
    "GCA_000007445.1\tB2-2\n",
    "\n",
    "GCA_000690815.1\tB2-2\n",
    "\n",
    "GCA_000734955.1\tB2-2\n",
    "\n",
    "GCA_000613265.1\tB2-2\n",
    "\n",
    "GCA_000333215.1\tB2-2\n",
    "\n",
    "GCA_000714595.1\tB2-2\n",
    "\n",
    "GCA_000167875.1\tB1\n",
    "\n",
    "GCA_000012025.1\tB1\n",
    "\n",
    "GCA_000005845.2\tA\n",
    "\n",
    "GCA_000010245.1\tA\n",
    "\n",
    "GCA_002007705.1\tA\n",
    "\n",
    "GCA_900092615.1\tA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Thursday June 27, 2019\n",
    "Today, Kaleb sent me a dataset that contains 14 *E. coli* genomes with one representative genome from each of the 14 phylotypes that he has classified. Each genome is the \"medioid\" of its phylotype, meaning it is the genome that has the least variability from all of the other genomes in its phylotype. These genomes should be a better test set to use for my program, and I am excited to see the eventual results. \n",
    "\n",
    "Genomes:\n",
    "\n",
    "Organism/Name,Strain,Assembly Accession,Size (Mb),TotalScore,SequenceScore,Bioproject,Biosample,Phylogroup\n",
    "Escherichia coli,MOD1-EC4358,GCA_002473875.1,5.34623,0.91,0.62,PRJNA230969,SAMN04913985,E2\n",
    "\n",
    "Escherichia coli,MOD1-EC1641,GCA_002460705.1,5.20103,0.88,0.62,PRJNA230969,SAMN05607397,E1\n",
    "\n",
    "Shigella flexneri,FC1170,GCA_002241955.1,4.59581,0.85,0.5,PRJNA393761,SAMN07340962,Shig1\n",
    "\n",
    "Shigella sonnei,CDPH_C96,GCA_001688385.1,4.44053,0.85,0.52,PRJNA300887,SAMN04440977,Shig2\n",
    "\n",
    "Escherichia coli,104,GCA_900096825.1,4.63813,0.99,0.95,PRJEB15352,SAMEA4444105,A\n",
    "\n",
    "Escherichia coli,MOD1-EC3802,GCA_002516075.1,4.61361,0.92,0.86,PRJNA230969,SAMN05440280,B1\n",
    "\n",
    "Escherichia coli,2014C-4638,GCA_003015295.1,5.77998,1,0.99,PRJNA218110,SAMN03265975,C\n",
    "\n",
    "Escherichia coli BIDMC 63,BIDMC 63,GCA_000692435.1,5.16234,0.99,0.97,PRJNA234151,SAMN02581284,B2-1\n",
    "\n",
    "Escherichia coli,GSK25213,GCA_000807565.1,4.94925,0.9,0.71,PRJNA221787,SAMN03169672,B2-2\n",
    "\n",
    "Escherichia coli FVEC1412,FVEC1412,GCA_000163235.1,5.23418,0.95,0.79,PRJNA39917,SAMN02595360,D1\n",
    "\n",
    "Escherichia coli,MOD1-EC5921,GCA_002534845.1,4.7781,0.95,0.88,PRJNA230969,SAMN05452849,D2\n",
    "\n",
    "Escherichia coli HVH 65 (4-2262045),HVH 65 (4-2262045),GCA_000456945.1,5.31618,0.97,0.9,PRJNA186093,SAMN01885720,D3\n",
    "\n",
    "Escherichia coli,WB61,GCA_002966755.1,4.68705,1,1,PRJNA374983,SAMN06341081,F\n",
    "\n",
    "Escherichia coli,MOD1-EC6621,GCA_002485345.1,4.90502,0.92,0.79,PRJNA230969,SAMN04992451,G\n",
    "\n",
    "### How to:\n",
    "* create a text file listing the accessions of all of the genomes\n",
    "* upload text file into [Batch Entrez](https://www.ncbi.nlm.nih.gov/sites/batchentrez), search for \"Assembly\"\n",
    "* download files from GenBank as Genomic FASTA files\n",
    "* unzip files from .fna.gz files to .fna files (click to open .fna.gz)\n",
    "* run prodigal to convert .fna files to .faa files (nucleotide sequence to amino acid sequence)\n",
    "* use kmerCounter.R to convert .faa files into kmer counts (amino acid sequence to kmer counts)\n",
    "* use multigenomeCSVGenerator.py to choose specific kmer counts and combine into one large matrix (kmer counts vectors to selected kmer count matrix)\n",
    "* use canopyClustering.py to group proteins (kmer count matrix into groups of proteins)\n",
    "* use grouping.R to visualize groups and get core/accessory/singleton composition for pangenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
